File: minimal-konjugaton-android/app/src/main/kotlin/com/konjugaton/hc/data/Loader.kt
/*
 * Load and validate the bundled taxonomy JSON, returning a pure-domain Catalog.
 *
 * Port of konjugaton's `data/loader.py` + `data/models.py` (German): @Serializable
 * DTOs are the I/O boundary (they mirror the JSON), then they're mapped to
 * framework-free domain objects — the domain never sees a serialization type.
 *
 * `parse(...)` takes the three JSON strings, so it is unit-testable on the plain
 * JVM. Reading the bytes from Android assets is the caller's job (AppState).
 */
package com.konjugaton.hc.data

import com.konjugaton.hc.domain.Auxiliary
import com.konjugaton.hc.domain.Catalog
import com.konjugaton.hc.domain.ConjugationData
import com.konjugaton.hc.domain.EndingTables
import com.konjugaton.hc.domain.SemanticContext
import com.konjugaton.hc.domain.Verb
import com.konjugaton.hc.domain.VerbClass
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private val JSON = Json { ignoreUnknownKeys = true }

// --- DTOs (mirror the asset JSON exactly) ---------------------------------

@Serializable
private data class ConjugationDto(
    @SerialName("praesens_stem_23") val praesensStem23: String? = null,
    @SerialName("praeteritum_stem") val praeteritumStem: String? = null,
    val partizip2: String? = null,
    @SerialName("konjunktiv2_stem") val konjunktiv2Stem: String? = null,
    val irregular: Map<String, Map<String, String>> = emptyMap(),
)

@Serializable
private data class VerbDto(
    val lemma: String,
    val translation: String,
    @SerialName("verb_class") val verbClass: String,
    val auxiliary: String,
    val transitive: Boolean,
    @SerialName("frequency_rank") val frequencyRank: Int,
    @SerialName("separable_prefix") val separablePrefix: String? = null,
    val family: String? = null,
    @SerialName("semantic_tags") val semanticTags: List<String> = emptyList(),
    val conjugation: ConjugationDto? = null,
)

@Serializable
private data class VerbsFileDto(val verbs: List<VerbDto>)

@Serializable
private data class ContextDto(
    val id: String,
    @SerialName("label_de") val labelDe: String,
    @SerialName("label_en") val labelEn: String,
    val templates: List<String>,
    val affinity: List<String> = emptyList(),
)

@Serializable
private data class ContextsFileDto(val contexts: List<ContextDto>)

// --- mapping DTO -> domain -------------------------------------------------

private fun VerbDto.toDomain(): Verb {
    val conj = conjugation
    return Verb(
        lemma = lemma,
        translation = translation,
        verbClass = VerbClass.fromValue(verbClass),
        auxiliary = Auxiliary.fromValue(auxiliary),
        transitive = transitive,
        frequencyRank = frequencyRank,
        conjugation = ConjugationData(
            praesensStem23 = conj?.praesensStem23,
            praeteritumStem = conj?.praeteritumStem,
            partizip2 = conj?.partizip2,
            konjunktiv2Stem = conj?.konjunktiv2Stem,
            irregular = conj?.irregular ?: emptyMap(),
        ),
        separablePrefix = separablePrefix,
        family = family,
        semanticTags = semanticTags,
    )
}

object CatalogLoader {
    /** Parse the three taxonomy JSON documents into a single [Catalog]. */
    fun parse(verbsJson: String, endingsJson: String, contextsJson: String): Catalog {
        val verbsFile = JSON.decodeFromString<VerbsFileDto>(verbsJson)
        val endingsRaw = JSON.decodeFromString<Map<String, Map<String, String>>>(endingsJson)
        val contextsFile = JSON.decodeFromString<ContextsFileDto>(contextsJson)

        val verbs = verbsFile.verbs.associate { it.lemma to it.toDomain() }
        val endingTables = EndingTables(endingsRaw)
        val contexts = contextsFile.contexts.associate {
            it.id to SemanticContext(
                id = it.id,
                labelDe = it.labelDe,
                labelEn = it.labelEn,
                templates = it.templates,
                affinity = it.affinity,
            )
        }

        return Catalog(verbs = verbs, endings = endingTables, contexts = contexts)
    }
}

File: src/konjugaton/data/loader.py
"""Load and validate the bundled data, returning a pure-domain :class:`Catalog`.

The catalog is the single read-only source of truth the engine queries. It is
cached, so the YAML is parsed and validated exactly once per process.
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from importlib import resources
from typing import TYPE_CHECKING, Any

import yaml

from konjugaton.data.models import ContextsFile, EndingsModel, VerbModel, VerbsFile
from konjugaton.domain import (
    ConjugationData,
    EndingTables,
    SemanticContext,
    Verb,
)

if TYPE_CHECKING:
    from collections.abc import Mapping

_DATA_PACKAGE = "konjugaton._data"

#: Paradigm names whose tables ship in endings.yaml.
_PARADIGMS: tuple[str, ...] = (
    "praesens",
    "praeteritum_weak",
    "praeteritum_strong",
    "konjunktiv",
)


class DataError(RuntimeError):
    """Bundled data could not be located or read.

    Almost always a *packaging* problem (the binary was built without bundling
    ``konjugaton._data``), not a logic bug — so the message says so loudly instead
    of surfacing an opaque ModuleNotFoundError from deep in the import machinery.
    """


def _read_yaml(filename: str) -> Any:
    try:
        resource = resources.files(_DATA_PACKAGE).joinpath(filename)
        text = resource.read_text(encoding="utf-8")
    except (ModuleNotFoundError, FileNotFoundError, OSError) as exc:
        raise DataError(
            f"Could not load bundled data '{filename}' from package "
            f"'{_DATA_PACKAGE}'. This is a packaging/bundling problem: the data "
            f"files were not shipped with the build. For a Nuitka binary, ensure "
            f"`--include-package-data=konjugaton` and that konjugaton/_data is a "
            f"package (has __init__.py). Original cause: {exc!r}"
        ) from exc
    return yaml.safe_load(text)


def _to_verb(model: VerbModel) -> Verb:
    conj = model.conjugation
    conjugation = ConjugationData(
        praesens_stem_23=conj.praesens_stem_23 if conj else None,
        praeteritum_stem=conj.praeteritum_stem if conj else None,
        partizip2=conj.partizip2 if conj else None,
        konjunktiv2_stem=conj.konjunktiv2_stem if conj else None,
        irregular={p: dict(slots) for p, slots in (conj.irregular if conj else {}).items()},
    )
    return Verb(
        lemma=model.lemma,
        translation=model.translation,
        verb_class=model.verb_class,
        auxiliary=model.auxiliary,
        transitive=model.transitive,
        frequency_rank=model.frequency_rank,
        conjugation=conjugation,
        separable_prefix=model.separable_prefix,
        family=model.family,
        semantic_tags=tuple(model.semantic_tags),
    )


@dataclass(frozen=True, slots=True)
class Catalog:
    """All loaded reference data, indexed for the engine."""

    verbs: Mapping[str, Verb]
    endings: EndingTables
    contexts: Mapping[str, SemanticContext]

    def verb(self, lemma: str) -> Verb:
        return self.verbs[lemma]

    @property
    def lemmas(self) -> list[str]:
        return list(self.verbs)

    @property
    def context_ids(self) -> list[str]:
        return list(self.contexts)


@lru_cache(maxsize=1)
def load_catalog() -> Catalog:
    """Parse, validate and index the bundled data (cached for the process)."""
    verbs_file = VerbsFile.model_validate(_read_yaml("verbs.yaml"))
    endings_model = EndingsModel.model_validate(_read_yaml("endings.yaml"))
    contexts_file = ContextsFile.model_validate(_read_yaml("contexts.yaml"))

    verbs = {m.lemma: _to_verb(m) for m in verbs_file.verbs}

    dump = endings_model.model_dump()
    ending_tables = EndingTables(tables={p: dict(dump[p]) for p in _PARADIGMS})

    contexts = {
        c.id: SemanticContext(
            id=c.id,
            label_de=c.label_de,
            label_en=c.label_en,
            templates=tuple(c.templates),
            affinity=tuple(c.affinity),
        )
        for c in contexts_file.contexts
    }

    return Catalog(verbs=verbs, endings=ending_tables, contexts=contexts)
